File: context_window_economics/settlement.py
"""Settlement engine for bilateral agent cost allocation.

Handles CMR verification, settlement proposal generation, settlement batching,
and provides the payment-rail abstraction interface (Section 13.2).
"""
import hashlib
import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Protocol, Tuple

from .allocation import allocate
from .schema import (
    AllocationMethod,
    CostMeteringRecord,
    DEFAULT_BATCH_THRESHOLD_USD,
    DEFAULT_BATCH_WINDOW_SECONDS,
    DEFAULT_SETTLEMENT_THRESHOLD,
    SettlementProposal,
    SettlementTier,
)


# ---------------------------------------------------------------------------
# Payment rail abstraction (Section 13.2)
# ---------------------------------------------------------------------------

class PaymentRail:
    """Abstract payment rail interface.

    Concrete implementations wrap x402, MPP, L402, Superfluid, or custom rails.
    The default implementation is a no-op ledger for metering-only deployments.
    """

    def commit_deposit(self, amount_usd: float, escrow_id: str) -> bool:
        """Lock funds in escrow."""
        return True

    def release_deposit(self, escrow_id: str, to_agent: str) -> bool:
        """Release escrowed funds to the specified agent."""
        return True

    def forfeit_deposit(self, escrow_id: str) -> bool:
        """Forfeit escrowed funds (spam penalty)."""
        return True

    def settle(
        self,
        from_agent: str,
        to_agent: str,
        amount_usd: float,
        interaction_id: str,
    ) -> "SettlementReceipt":
        """Execute a one-time settlement payment."""
        return SettlementReceipt(
            interaction_id=interaction_id,
            from_agent=from_agent,
            to_agent=to_agent,
            amount_usd=amount_usd,
            status="completed",
            timestamp=datetime.now(timezone.utc).isoformat(),
        )

    def stream_open(
        self,
        from_agent: str,
        to_agent: str,
        rate_usd_per_second: float,
    ) -> str:
        """Open a streaming payment channel. Returns a stream handle."""
        return f"stream-{from_agent}-{to_agent}"

    def stream_close(self, handle: str) -> "SettlementReceipt":
        """Close a streaming payment channel."""
        return SettlementReceipt(
            interaction_id=handle,
            from_agent="",
            to_agent="",
            amount_usd=0.0,
            status="closed",
            timestamp=datetime.now(timezone.utc).isoformat(),
        )


@dataclass
class SettlementReceipt:
    """Receipt from a completed settlement."""
    interaction_id: str = ""
    from_agent: str = ""
    to_agent: str = ""
    amount_usd: float = 0.0
    status: str = "completed"
    timestamp: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    tx_ref: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "interaction_id": self.interaction_id,
            "from_agent": self.from_agent,
            "to_agent": self.to_agent,
            "amount_usd": self.amount_usd,
            "status": self.status,
            "timestamp": self.timestamp,
            "tx_ref": self.tx_ref,
        }


# ---------------------------------------------------------------------------
# Settlement Engine
# ---------------------------------------------------------------------------

class SettlementEngine:
    """Core settlement engine that evaluates CMRs and produces settlement proposals.

    Supports all three CWEP tiers:
      - Tier 1: Metering only (no settlement)
      - Tier 2: Rule-based (static allocation rules)
      - Tier 3: Dynamic (Shapley/Nash)
    """

    def __init__(
        self,
        tier: str = SettlementTier.TIER_1_METERING.value,
        method: str = AllocationMethod.SHAPLEY.value,
        threshold_usd: float = DEFAULT_SETTLEMENT_THRESHOLD,
        payment_rail: Optional[PaymentRail] = None,
        tolerance: float = 0.05,
        **allocation_kwargs,
    ):
        """
        Args:
            tier: Settlement tier.
            method: Default allocation method for Tier 2/3.
            threshold_usd: Minimum cost to trigger settlement.
            payment_rail: Payment rail implementation (default: no-op).
            tolerance: Agreement tolerance as fraction of total cost (default 5%).
            **allocation_kwargs: Extra params passed to allocation methods.
        """
        self.tier = tier
        self.method = method
        self.threshold_usd = threshold_usd
        self.payment_rail = payment_rail or PaymentRail()
        self.tolerance = tolerance
        self.allocation_kwargs = allocation_kwargs

    def propose(self, cmr: CostMeteringRecord) -> Optional[SettlementProposal]:
        """Generate a settlement proposal for the given CMR.

        Returns None for Tier 1 or if total cost is below threshold.
        """
        if self.tier == SettlementTier.TIER_1_METERING.value:
            return None

        if cmr.totals.total_cost_usd < self.threshold_usd:
            return None

        return allocate(cmr, method=self.method, **self.allocation_kwargs)

    def settle(
        self, cmr: CostMeteringRecord, proposal: Optional[SettlementProposal] = None
    ) -> Optional[SettlementReceipt]:
        """Execute settlement for a CMR.

        If no proposal is provided, one is generated. If the proposal's net
        transfer is zero or below threshold, no payment is executed.
        """
        if proposal is None:
            proposal = self.propose(cmr)
        if proposal is None:
            return None

        if proposal.net_transfer_usd < self.threshold_usd:
            return None

        # Attach proposal to CMR
        cmr.settlement = proposal

        # Determine from/to
        if proposal.transfer_direction == "requestor_to_responder":
            from_agent = cmr.requestor.agent_id if cmr.requestor else ""
            to_agent = cmr.responder.agent_id if cmr.responder else ""
        else:
            from_agent = cmr.responder.agent_id if cmr.responder else ""
            to_agent = cmr.requestor.agent_id if cmr.requestor else ""

        return self.payment_rail.settle(
            from_agent=from_agent,
            to_agent=to_agent,
            amount_usd=proposal.net_transfer_usd,
            interaction_id=cmr.interaction_id,
        )

    def verify_proposals(
        self,
        proposal_a: SettlementProposal,
        proposal_b: SettlementProposal,
        total_cost: float,
    ) -> Tuple[bool, float]:
        """Verify two agents' settlement proposals agree within tolerance.

        Returns (agreed: bool, discrepancy: float).
        """
        discrepancy = abs(
            proposal_a.requestor_pays_usd - proposal_b.requestor_pays_usd
        )
        threshold = total_cost * self.tolerance
        return discrepancy <= threshold, discrepancy


# ---------------------------------------------------------------------------
# Settlement Batching (Section 13.3)
# ---------------------------------------------------------------------------

class SettlementBatch:
    """Accumulates CMRs and computes net settlement across a batch window.

    Reduces transaction fees by settling the net amount across many interactions.
    """

    def __init__(
        self,
        window_seconds: int = DEFAULT_BATCH_WINDOW_SECONDS,
        threshold_usd: float = DEFAULT_BATCH_THRESHOLD_USD,
        method: str = AllocationMethod.SHAPLEY.value,
        **allocation_kwargs,
    ):
        self.window_seconds = window_seconds
        self.threshold_usd = threshold_usd
        self.method = method
        self.allocation_kwargs = allocation_kwargs
        self._cmrs: List[CostMeteringRecord] = []
        self._started: Optional[datetime] = None

    def add(self, cmr: CostMeteringRecord) -> None:
        """Add a CMR to the current batch."""
        if self._started is None:
            self._started = datetime.now(timezone.utc)
        self._cmrs.append(cmr)

    @property
    def count(self) -> int:
        return len(self._cmrs)

    def should_flush(self) -> bool:
        """Check if the batch should be flushed (time or $ threshold)."""
        if not self._cmrs:
            return False

        # Check time window
        if self._started is not None:
            elapsed = (datetime.now(timezone.utc) - self._started).total_seconds()
            if elapsed >= self.window_seconds:
                return True

        # Check net amount threshold
        net = self._compute_net()
        return abs(net) >= self.threshold_usd

    def flush(self) -> Dict[str, Any]:
        """Compute net settlement for the batch and reset.

        Returns:
            Dict with agent_pair, net_amount, direction, interaction_count, cmr_ids.
        """
        if not self._cmrs:
            return {
                "net_amount_usd": 0.0,
                "direction": "",
                "interaction_count": 0,
                "cmr_ids": [],
            }

        net = self._compute_net()
        cmr_ids = [c.interaction_id for c in self._cmrs]
        count = len(self._cmrs)

        if net >= 0:
            direction = "requestor_to_responder"
        else:
            direction = "responder_to_requestor"
            net = -net

        result = {
            "net_amount_usd": net,
            "direction": direction,
            "interaction_count": count,
            "cmr_ids": cmr_ids,
        }

        # Reset
        self._cmrs = []
        self._started = None
        return result

    def _compute_net(self) -> float:
        """Compute net transfer across all CMRs in the batch.

        Positive = requestor owes responder.
        """
        total_net = 0.0
        for cmr in self._cmrs:
            proposal = allocate(cmr, method=self.method, **self.allocation_kwargs)
            if proposal.transfer_direction == "requestor_to_responder":
                total_net += proposal.net_transfer_usd
            else:
                total_net -= proposal.net_transfer_usd
        return total_net


# ---------------------------------------------------------------------------
# CMR verification utilities
# ---------------------------------------------------------------------------

def cmr_hash(cmr: CostMeteringRecord) -> str:
    """Compute SHA-256 hash of a CMR for verification/exchange."""
    data = json.dumps(cmr.to_dict(), sort_keys=True).encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def verify_cmr_pair(cmr_a: CostMeteringRecord, cmr_b: CostMeteringRecord) -> bool:
    """Verify that two independently generated CMRs describe the same interaction.

    Checks interaction_id and token flow agreement.
    """
    if cmr_a.interaction_id != cmr_b.interaction_id:
        return False

    for flow_key in cmr_a.flows:
        if flow_key not in cmr_b.flows:
            return False
        if cmr_a.flows[flow_key].tokens != cmr_b.flows[flow_key].tokens:
            return False
    return True

File: typescript/src/settlement.ts
import { createHash } from "node:crypto";
import { allocate } from "./allocation";
import {
  AllocationMethod,
  CostMeteringRecord,
  DEFAULT_BATCH_THRESHOLD_USD,
  DEFAULT_BATCH_WINDOW_SECONDS,
  DEFAULT_SETTLEMENT_THRESHOLD,
  SettlementProposal,
  SettlementTier,
} from "./types";

// ---------------------------------------------------------------------------
// Payment rail abstraction
// ---------------------------------------------------------------------------

export interface SettlementReceiptData {
  interaction_id: string;
  from_agent: string;
  to_agent: string;
  amount_usd: number;
  status: string;
  timestamp: string;
  tx_ref: string | null;
}

export class SettlementReceipt {
  interactionId: string;
  fromAgent: string;
  toAgent: string;
  amountUsd: number;
  status: string;
  timestamp: string;
  txRef: string | null;

  constructor(opts?: Partial<{
    interactionId: string;
    fromAgent: string;
    toAgent: string;
    amountUsd: number;
    status: string;
    timestamp: string;
    txRef: string | null;
  }>) {
    this.interactionId = opts?.interactionId ?? "";
    this.fromAgent = opts?.fromAgent ?? "";
    this.toAgent = opts?.toAgent ?? "";
    this.amountUsd = opts?.amountUsd ?? 0.0;
    this.status = opts?.status ?? "completed";
    this.timestamp = opts?.timestamp ?? new Date().toISOString();
    this.txRef = opts?.txRef ?? null;
  }

  toDict(): SettlementReceiptData {
    return {
      interaction_id: this.interactionId,
      from_agent: this.fromAgent,
      to_agent: this.toAgent,
      amount_usd: this.amountUsd,
      status: this.status,
      timestamp: this.timestamp,
      tx_ref: this.txRef,
    };
  }
}

export class PaymentRail {
  commitDeposit(_amountUsd: number, _escrowId: string): boolean {
    return true;
  }

  releaseDeposit(_escrowId: string, _toAgent: string): boolean {
    return true;
  }

  forfeitDeposit(_escrowId: string): boolean {
    return true;
  }

  settle(
    fromAgent: string,
    toAgent: string,
    amountUsd: number,
    interactionId: string
  ): SettlementReceipt {
    return new SettlementReceipt({
      interactionId,
      fromAgent,
      toAgent,
      amountUsd,
      status: "completed",
      timestamp: new Date().toISOString(),
    });
  }

  streamOpen(fromAgent: string, toAgent: string, _rateUsdPerSecond: number): string {
    return `stream-${fromAgent}-${toAgent}`;
  }

  streamClose(handle: string): SettlementReceipt {
    return new SettlementReceipt({
      interactionId: handle,
      status: "closed",
      timestamp: new Date().toISOString(),
    });
  }
}

// ---------------------------------------------------------------------------
// Settlement Engine
// ---------------------------------------------------------------------------

export class SettlementEngine {
  readonly tier: string;
  readonly method: string;
  readonly thresholdUsd: number;
  readonly paymentRail: PaymentRail;
  readonly tolerance: number;
  readonly allocationKwargs: Record<string, number>;

  constructor(opts?: Partial<{
    tier: string;
    method: string;
    thresholdUsd: number;
    paymentRail: PaymentRail;
    tolerance: number;
    allocationKwargs: Record<string, number>;
  }>) {
    this.tier = opts?.tier ?? SettlementTier.TIER_1_METERING;
    this.method = opts?.method ?? AllocationMethod.SHAPLEY;
    this.thresholdUsd = opts?.thresholdUsd ?? DEFAULT_SETTLEMENT_THRESHOLD;
    this.paymentRail = opts?.paymentRail ?? new PaymentRail();
    this.tolerance = opts?.tolerance ?? 0.05;
    this.allocationKwargs = opts?.allocationKwargs ?? {};
  }

  propose(cmr: CostMeteringRecord): SettlementProposal | null {
    if (this.tier === SettlementTier.TIER_1_METERING) return null;
    if (cmr.totals.totalCostUsd < this.thresholdUsd) return null;
    return allocate(cmr, this.method, this.allocationKwargs);
  }

  settle(
    cmr: CostMeteringRecord,
    proposal?: SettlementProposal | null
  ): SettlementReceipt | null {
    if (proposal === undefined || proposal === null) {
      proposal = this.propose(cmr);
    }
    if (!proposal) return null;
    if (proposal.netTransferUsd < this.thresholdUsd) return null;

    cmr.settlement = proposal;

    let fromAgent: string;
    let toAgent: string;
    if (proposal.transferDirection === "requestor_to_responder") {
      fromAgent = cmr.requestor?.agentId ?? "";
      toAgent = cmr.responder?.agentId ?? "";
    } else {
      fromAgent = cmr.responder?.agentId ?? "";
      toAgent = cmr.requestor?.agentId ?? "";
    }

    return this.paymentRail.settle(
      fromAgent,
      toAgent,
      proposal.netTransferUsd,
      cmr.interactionId
    );
  }

  verifyProposals(
    proposalA: SettlementProposal,
    proposalB: SettlementProposal,
    totalCost: number
  ): [boolean, number] {
    const discrepancy = Math.abs(
      proposalA.requestorPaysUsd - proposalB.requestorPaysUsd
    );
    const threshold = totalCost * this.tolerance;
    return [discrepancy <= threshold, discrepancy];
  }
}

// ---------------------------------------------------------------------------
// Settlement Batching
// ---------------------------------------------------------------------------

export class SettlementBatch {
  readonly windowSeconds: number;
  readonly thresholdUsd: number;
  readonly method: string;
  readonly allocationKwargs: Record<string, number>;
  private _cmrs: CostMeteringRecord[] = [];
  private _started: Date | null = null;

  constructor(opts?: Partial<{
    windowSeconds: number;
    thresholdUsd: number;
    method: string;
    allocationKwargs: Record<string, number>;
  }>) {
    this.windowSeconds = opts?.windowSeconds ?? DEFAULT_BATCH_WINDOW_SECONDS;
    this.thresholdUsd = opts?.thresholdUsd ?? DEFAULT_BATCH_THRESHOLD_USD;
    this.method = opts?.method ?? AllocationMethod.SHAPLEY;
    this.allocationKwargs = opts?.allocationKwargs ?? {};
  }

  add(cmr: CostMeteringRecord): void {
    if (this._started === null) {
      this._started = new Date();
    }
    this._cmrs.push(cmr);
  }

  get count(): number {
    return this._cmrs.length;
  }

  shouldFlush(): boolean {
    if (this._cmrs.length === 0) return false;

    if (this._started !== null) {
      const elapsed = (Date.now() - this._started.getTime()) / 1000;
      if (elapsed >= this.windowSeconds) return true;
    }

    const net = this._computeNet();
    return Math.abs(net) >= this.thresholdUsd;
  }

  flush(): Record<string, unknown> {
    if (this._cmrs.length === 0) {
      return {
        net_amount_usd: 0.0,
        direction: "",
        interaction_count: 0,
        cmr_ids: [],
      };
    }

    let net = this._computeNet();
    const cmrIds = this._cmrs.map((c) => c.interactionId);
    const count = this._cmrs.length;

    let direction: string;
    if (net >= 0) {
      direction = "requestor_to_responder";
    } else {
      direction = "responder_to_requestor";
      net = -net;
    }

    const result = {
      net_amount_usd: net,
      direction,
      interaction_count: count,
      cmr_ids: cmrIds,
    };

    this._cmrs = [];
    this._started = null;
    return result;
  }

  private _computeNet(): number {
    let totalNet = 0.0;
    for (const cmr of this._cmrs) {
      const proposal = allocate(cmr, this.method, this.allocationKwargs);
      if (proposal.transferDirection === "requestor_to_responder") {
        totalNet += proposal.netTransferUsd;
      } else {
        totalNet -= proposal.netTransferUsd;
      }
    }
    return totalNet;
  }
}

// ---------------------------------------------------------------------------
// CMR verification utilities
// ---------------------------------------------------------------------------

function sortedStringify(value: unknown): string {
  if (value === null || value === undefined) return JSON.stringify(value);
  if (Array.isArray(value)) {
    return "[" + value.map(sortedStringify).join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    return "{" + keys.map(k => JSON.stringify(k) + ":" + sortedStringify(obj[k])).join(",") + "}";
  }
  return JSON.stringify(value);
}

export function cmrHash(cmr: CostMeteringRecord): string {
  const data = sortedStringify(cmr.toDict());
  return createHash("sha256").update(data, "utf-8").digest("hex");
}

export function verifyCmrPair(
  cmrA: CostMeteringRecord,
  cmrB: CostMeteringRecord
): boolean {
  if (cmrA.interactionId !== cmrB.interactionId) return false;

  for (const flowKey of Object.keys(cmrA.flows)) {
    if (!(flowKey in cmrB.flows)) return false;
    if (cmrA.flows[flowKey].tokens !== cmrB.flows[flowKey].tokens) return false;
  }
  return true;
}
